---
title: "GAND - Single Cell Analysis Report"
author: "Molly Easter, Patrick CN Martin"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output:
  pdf_document:
    toc: true
    toc_depth: 3
    number_sections: true
params:
    input_dir: ""
---

```{r setup_and_libs, eval = TRUE, echo = FALSE}
# LIBS
library(future)
library(Matrix)
library(Seurat)
library(dplyr)
library(tidyr)
library(patchwork)
library(DESeq2)
library(ggplot2)
library(rmarkdown)
library(hdf5r)
library(ggpubr)

set.seed(42)
# ARGPARSER
input_dir <- params$input_dir

# FUTURE OPTIONS
max_size <- 100000 * 1024^2
options(future.globals.maxSize = max_size)
```

# Single Cell RNA-seq Analysis of Gatad2b
This report highlight the main findings related to the analysis of single cell RNA-seq of Gatad2b
deficient samples.

The full analysis is available on (GitHub)[https://github.com/patrickCNMartin/GAND] as
a Nextflow pipeline. Containerization of the analysis will be made availble in the comming weeks.

NOTE: All figures are currently being generated and saved through this report.

## scRNA-seq Experiment

We performed single-cell RNA sequencing (scRNA-seq) on cortical samples obtained 
from Gatad2bstop/+ and WT littermates at a single embryonic time point (E16.5) in order 
to explore underlying mechanisms of Gatad2b function. We also generated Gatad2bstop/stop 
scRNA-seq data to identify dosage sensitive targets of Gatad2b.

The data is availble under accession number (GSE244477)[https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE244477]

```{r indiv_umaps, eval = TRUE, echo = FALSE}
preprocessed <- readRDS(params$input_dir)
umap_plots <- vector("list", length(preprocessed))
for (i in seq_along(preprocessed)) {
    tmp <- preprocessed[[i]]
    tag <- paste(unique(tmp@meta.data$sample),"-",unique(tmp@meta.data$condition))
    umap_plots[[i]] <- DimPlot(tmp, reduction = "umap") +
                       ggtitle(paste0("UMAP for ", tag))
}
grid_plots <- wrap_plots(umap_plots, ncol = 3)
```

```{r indiv_umap_out, eval = TRUE, echo = FALSE, fig.width = 15, fig.height = 15,
    fig.caption = "UMAP plots for each Individual data set."}
print(grid_plots)
pdf("figure_1_individual_umap_plots.pdf",width = 15, height = 15)
print(grid_plots)
dev.off()
```

## scRNA-seq Reference Data

To annotate the scRNA data, we select a reference data set with pre-annotated cells.
The reference data set is availbale under accession number (GSE123335)[https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE123335]
```